gensim/models/ensemblelda.py

```diff
@@ -0,0 +1,1308 @@
+                "label": label,
+                "num_cores": len([topic for topic in group if topic["is_core"]]),
+            })
+
+        logger.info("found %s clusters", len(clusters))
+
+        return clusters
+
+    def _remove_from_all_sets(self, label, clusters):
+        """Remove a label from every set in "neighboring_labels" for each core in ``clusters``."""
+        for cluster in clusters:
+            for neighboring_labels_set in cluster["neighboring_labels"]:
+                if label in neighboring_labels_set:
+                    neighboring_labels_set.remove(label)
+
+    def _contains_isolated_cores(self, label, cluster, min_cores):
+        """Check if the cluster has at least ``min_cores`` of cores that belong to no other cluster."""
+        return sum(map(lambda x: x == {label}, cluster["neighboring_labels"])) >= min_cores
+
+    def _generate_stable_topics(self, min_cores=None):
+        """Generate stable topics out of the clusters.
+
+        The function finds clusters of topics using a variant of DBScan.  If a cluster has enough core topics
+        (c.f. parameter ``min_cores``), then this cluster represents a stable topic.  The stable topic is specifically
+        calculated as the average over all topic-term distributions of the core topics in the cluster.
+
+        This function is the last step that has to be done in the ensemble.  After this step is complete,
+        Stable topics can be retrieved afterwards using the :meth:`~gensim.models.ensemblelda.EnsembleLda.get_topics`
+        method.
+
+        Parameters
+        ----------
+        min_cores : int
+            Minimum number of core topics needed to form a cluster that represents a stable topic.
+                Using ``None`` defaults to ``min_cores = min(3, max(1, int(self.num_models /4 +1)))``
+
+        """
+        # min_cores being 0 makes no sense. there has to be a core for a cluster
+        # or there is no cluster
+        if min_cores == 0:
+            min_cores = 1
+
+        if min_cores is None:
+            # min_cores is a number between 1 and 3, depending on the number of models
+            min_cores = min(3, max(1, int(self.num_models / 4 + 1)))
+            logger.info("generating stable topics, using %s for min_cores", min_cores)
+        else:
+            logger.info("generating stable topics")
+
+        results = self.cluster_model.results
+
+        grouped_by_labels = self._group_by_labels(results)
+
+        clusters = self._aggregate_topics(grouped_by_labels)
+
+        # Clusters with noisy invalid neighbors may have a harder time being marked as stable, so start with the
+        # easy ones and potentially already remove some noise by also sorting smaller clusters to the front.
+        # This clears up the clusters a bit before checking the ones with many neighbors.
+        sorted_clusters = sorted(
+            clusters,
+            key=lambda cluster: (
```

Comment:
Are there any benefits in using a lambda function over a normal function here? If not, I recommend using a pure function for readability.

```python
def _cluster_sort_key(cluster):
    return (cluster["max_num_neighboring_labels"], cluster["num_cores"], cluster["label"])

sorted_clusters = sort(clusters, key=cluster_sort_key, reverse=False)
```

https://treyhunner.com/2018/09/stop-writing-lambda-expressions/#The_pros_and_cons_of_lambda

Reply:
> I find the name of that key function makes it clearer what we’re sorting by. We’re not just sorting by the length and we’re not just sorting by the color: we’re sorting by both.

Knowing what the key parameter does for the sort method is sufficient to understand that. I don't think python-internals require any explanation, docstrings or comments, they require looking up pythons documentation and learning how sorting in python works. I also don't think our sort call is hard to read, I think it is very clear what we are sorting by. Moving it to a separate function means we have to jump there to see that very trivial construction of a tuple and then jump back, which I think is worse.

Reply:
I just want to kindly jump in 😊

@mpenkov , it's a good point.  GIven the function has the following attributes:

- one line
- no docstring
- is not used elsewhere
- no expressive name that is more succinct than the function code itself

it seems to me to be more readable as a lambda.  Six of one, half-dosen of the other...  I changed it, sorry @sezanzeb 😉